File: Python3/3033_modify_the_matrix.py
# https://leetcode.cn/problems/modify-the-matrix/

class Solution:
    def modifiedMatrix(self, matrix):
        """
        Inputs:
            matrix: list[list[int]]
        Outputs:
            res: list[list[int]]
        """
        m, n = len(matrix), len(matrix[0])
        max_col = [0 for _ in range(n)]
        replace = []
        for j in range(n):
            max_val = matrix[0][j]
            for i in range(m):
                val = matrix[i][j]
                max_val = max(max_val, val)
                if val == -1:
                    replace.append([i,j])
            max_col[j] = max_val
            
        for i, j in replace:
            matrix[i][j] = max_col[j]
        return matrix

File: Cpp/3033_modify_the_matrix.cpp
// https://leetcode.cn/problems/modify-the-matrix/
        
class Solution {
public:
    vector<vector<int>> modifiedMatrix(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<int> max_col(n);
        vector<pair<int, int>> replace;
        for (int j = 0; j < n; ++j) {
            int max_val = matrix[0][j];
            for (int i = 0; i < m; ++i) {
                int val = matrix[i][j];
                max_val = max(max_val, val);
                if (val == -1) {
                    replace.push_back(make_pair(i,j));
                }
            }
            max_col[j] = max_val;
        }
        for (auto &idx: replace) {
            matrix[idx.first][idx.second] = max_col[idx.second];
        }
        return matrix;
    }
};
